src/Concordium/Client/GRPC2.hs

```diff
@@ -0,0 +1,2693 @@
+instance ToProto IpPort where
+    type Output IpPort = Proto.Port
+    toProto ip = Proto.make $ ProtoFields.value .= fromIntegral ip
+
+instance FromProto Proto.IpAddress where
+    type Output' Proto.IpAddress = IpAddress
+    fromProto peer = peer ^? ProtoFields.value
+
+instance FromProto Proto.Port where
+    type Output' Proto.Port = IpPort
+    fromProto port = fromIntegral <$> port ^? ProtoFields.value
+
+instance FromProto Proto.BannedPeer where
+    type Output' Proto.BannedPeer = Peer
+    fromProto peer = fromProto =<< peer ^? ProtoFields.ipAddress
+
+instance FromProto Proto.BannedPeers where
+    type Output' Proto.BannedPeers = [Peer]
+    fromProto peers = mapM fromProto =<< peers ^? ProtoFields.peers
+
+instance FromProto Proto.InstanceStateValueAtKey where
+    type Output' Proto.InstanceStateValueAtKey = ByteString
+    fromProto bs = bs ^? ProtoFields.value
+
+instance FromProto Proto.Signature where
+    type Output' Proto.Signature = Signature
+    fromProto sig = Signature . BSS.toShort <$> (sig ^? ProtoFields.value)
+
+instance FromProto Proto.SignatureMap where
+    type Output' Proto.SignatureMap = Updates.UpdateInstructionSignatures
+    fromProto uis = do
+        signatures <-
+            fmap Map.fromAscList
+                . mapM deMk
+                . Map.toAscList
+                =<< uis ^? ProtoFields.signatures
+        return Updates.UpdateInstructionSignatures{..}
+      where
+        deMk (k, s) = do
+            sig <- fromProto s
+            return (fromIntegral k, sig)
+
+instance FromProto Proto.TransactionTime where
+    type Output' Proto.TransactionTime = TransactionTime
+    fromProto = return . TransactionTime . deMkWord64
+
+instance FromProto Proto.UpdateSequenceNumber where
+    type Output' Proto.UpdateSequenceNumber = Nonce
+    fromProto = return . Nonce . deMkWord64
+
+instance FromProto Proto.UpdateInstructionHeader where
+    type Output' Proto.UpdateInstructionHeader = Updates.UpdateHeader
+    fromProto uiHeader = do
+        updateSeqNumber <- fromProto =<< uiHeader ^? ProtoFields.sequenceNumber
+        updateEffectiveTime <- fromProto =<< uiHeader ^? ProtoFields.effectiveTime
+        updateTimeout <- fromProto =<< uiHeader ^? ProtoFields.timeout
+        let updatePayloadSize = PayloadSize 0 -- FIXME: What should go here, no such field in the protobuf equivalent.
+        return Updates.UpdateHeader{..}
+
+instance FromProto Proto.UpdateInstruction where
+    type Output' Proto.UpdateInstruction = Updates.UpdateInstruction
+    fromProto ui = do
+        uiSignHash <- Nothing -- FIXME: What should go here, no such field in the protobuf equivalent.
+        uiSignatures <- fromProto =<< ui ^? ProtoFields.signatures
+        uiHeader <- fromProto =<< ui ^? ProtoFields.header
+        uiPayload <- do
+            pl <- ui ^? (ProtoFields.payload . ProtoFields.rawPayload)
+            case (S.runGet $ Updates.getUpdatePayload SP1) pl of -- FIXME: What protocol version to use here?
```

Comment:
So if the payload is decoded w.r.t a specific protocol version, it sounds like we should add a field for either the PV or the undecoded payload in protocol buffer schema and include that in the message from the node side of things?

Reply:
This is just an artifact of how it is modelled in Haskell.

The simplest solution is to add a version of UpdateInstruction that has undecoded payload.

The only other option would be that you know the protocol version here. That you could technically do but it will be messy and involve boilerplate code.

Since this endpoint does not exist for V1, and is thus currently not used by anything in concordium-client, please put it on hold. We will handle it in some followup work. I don't want to delay migration of the client any further because of this.